### ISSUE TITLE
Nix and Poetry `kmir run` have different output

### ISSUE THREAD

**Issue by dkcumming** [runtimeverification/mir-semantics#200]
### Description
I discovered that there is a difference in output between `kmir run` depending on if kmir is built with `nix` or `poetry`. The difference is in the production of `</generatedTop>` closing cell, and `<generatedCounter>` cells and value. Below is the `vimdiff` that clearly displays the difference.

### Reproducer
```
mir-semantics$ git branch
* (HEAD detached at origin/master)
mir-semantics$ cd kmir/
kmir$ nix shell
path '/home/daniel/Applications/mir-semantics/kmir' does not contain a 'flake.nix', searching up
kmir$ which kmir
/nix/store/wafzmd218f9vzdy399h8bccph048zp0i-python3.10-kmir-0.1.0/bin/kmir
kmir$ kmir run --output pretty src/tests/integration/test-data/handwritten-mir/execution/arithm-simple.mir > nix-default.mir
kmir$ 
exit
kmir$ poetry shell
Spawning shell within /home/daniel/.cache/pypoetry/virtualenvs/kmir-MbM7GIng-py3.10
kmir$ . /home/daniel/.cache/pypoetry/virtualenvs/kmir-MbM7GIng-py3.10/bin/activate
kmir$ which kmir
/home/daniel/.cache/pypoetry/virtualenvs/kmir-MbM7GIng-py3.10/bin/kmir
kmir$ kbuild which llvm
/home/daniel/.kbuild/mir-semantics/5a0ac5c/target/v6.0.63/llvm
kmir$ kmir run --definition-dir $(kbuild which llvm) --output pretty src/tests/integration/test-data/handwritten-mir/execution/arithm-simple.mir > poetry.mir
kmir$ vimdiff nix-default.mir poetry.mir
2 files to edit
kmir$ 
```

The vimdiff output:
Note that the first line is actually not a problem, that line should of been piped to `stderr` which I will fix later. The bottom lines are the problem.
![nix-poetry-error](https://github.com/runtimeverification/mir-semantics/assets/124537596/a339503b-0ee8-41b4-a740-092916fa0db2)

### Solution:
Unknown at this point, I need to investigate. But potentially there is a problem with the difference in how we build the parser with nix as opposed to poetry. Note [this line](https://github.com/runtimeverification/mir-semantics/blob/4dfdd16a87805735e36b5f04c7881c97f381c472/flake.nix#L68) in `flake.nix`, I don't have any idea why it would be a problem but it is the first thing that is suspicious in my mind.

**Comment by geo2a** [comment]
## **TL;DR**

The reason this happens has nothing to do with `poetry`, but rather with which kompiled definition is being used by `kmir`. The two calls above definitions kompiled with different version of K.

Using K v6.0.69 gives us the latter result. There is probably a bug in the pretty-printer, since the configuration should look something like
```
<generatedTop>
  <kmir> ... </kmir>
  <generatedCounter> ... </generatedCounter>
</generatedTop>
```

But with the K v6.0.69 the unparsed result is

```
<generatedTop>
  <kmir> ... </kmir>
<generatedCounter>
</generatedCounter>
```

which is obviously wrong

## Longer explanation

In the reproducer snipped above, kmir run is invoked twice, once withing `nix shell` and the second time within `poetry shell`. To the best of my understanding, the Python code of `kmir` is the same in both invocations, however the compiled K definition is not. Let's look at both invocations using the `--verbose` flag I've added in #203. Using `--verbose` will output the calls to `krun` that `pyk` makes for us under the hood, making crystal cleat which definition is used.

### Nix-wrapped `kmir` 

It is very important that the Nix-wrapped `kmir` does not need `--definition-dir`, since the Nix flake already provides the `KMIR_LLVM_DIR` environment variable for it. This means that the kompiled definition built and placed in the Nix store will be used by Nix-wrapped `kmir` by default.

```
$ kmir run --verbose --output pretty src/tests/integration/test-data/handwritten-mir/execution/arithm-simple.mir > t2.pretty
INFO 2023-08-25 10:58:59,597 pyk.ktool.krun - Running: krun /tmp/tmpvon6ybq0 --definition /nix/store/ns15yr0bqbj40pplripjp8dccsqw8qjn-mir-semantics-049c7255dd6d8e63317440ec89627ac838e0db80/lib/llvm --output pretty -pPGM=/nix/store/ns15yr0bqbj40pplripjp8dccsqw8qjn-mir-semantics-049c7255dd6d8e63317440ec89627ac838e0db80/lib/llvm/parser_Mir_MIR-PARSER-SYNTAX
INFO 2023-08-25 10:58:59,894 pyk.ktool.krun - Completed in 0.297s with status 0: krun /tmp/tmpvon6ybq0 --definition /nix/store/ns15yr0bqbj40pplripjp8dccsqw8qjn-mir-semantics-049c7255dd6d8e63317440ec89627ac838e0db80/lib/llvm --output pretty -pPGM=/nix/store/ns15yr0bqbj40pplripjp8dccsqw8qjn-mir-semantics-049c7255dd6d8e63317440ec89627ac838e0db80/lib/llvm/parser_Mir_MIR-PARSER-SYNTAX
``` 
We see two INFO log messages: the first one is the call to `krun` and the second one is the status code of that call and the timing. Looking closer at the first one, we can see that we indeed use the kompiled definition from the Nix store. It is *very important* that this definition will have been built with the version of `kompile` specifying on `flake.nix`.

 ### Native `kmir`, outside Nix, i.e. in `poetry shell` or `pip`-installed 

When using `kmir` outside of Nix, we have to somehow give it the kompiled definition. We can either use the `--definition-dir` option or set `KMIR_LLVM_DIR`.

```
poetry run kmir run --debug --definition-dir $(kbuild which llvm) --output pretty src/tests/integration/test-data/handwritten-mir/execution/arithm-simple.mir
WARNING 2023-08-25 11:07:56,086 kmir.kmir - Haskell defintion could not be found
INFO 2023-08-25 11:07:56,087 pyk.ktool.krun - Running: krun /tmp/tmplfdw0bwl --definition /home/geo2a/.kbuild/mir-semantics/2679761/target/v6.0.71/llvm --output pretty -pPGM=/home/geo2a/.kbuild/mir-semantics/2679761/target/v6.0.71/llvm/parser_Mir_MIR-PARSER-SYNTAX
INFO 2023-08-25 11:07:56,281 pyk.ktool.krun - Completed in 0.194s with status 0: krun /tmp/tmplfdw0bwl --definition /home/geo2a/.kbuild/mir-semantics/2679761/target/v6.0.71/llvm --output pretty -pPGM=/home/geo2a/.kbuild/mir-semantics/2679761/target/v6.0.71/llvm/parser_Mir_MIR-PARSER-SYNTAX
```

Here, we're using the kompiled definition that K built gave us. We can also use the one built by Nix if we really want to.
